src/rsz/src/RepairDesign.cc

```diff
@@ -1791,86 +1810,223 @@ void RepairDesign::makeRepeater(const char* reason,
+        = db_network_->getNetDriverParentModule(load_net, driver_pin, true);
+    if (parent_module) {
+      odb::dbModInst* parent_mod_inst = parent_module->getModInst();
+      if (parent_mod_inst) {
+        parent = db_network_->dbToSta(parent_mod_inst);
+      } else {
+        parent = db_network_->topInstance();
+      }
+    } else {
+      parent = db_network_->topInstance();
+    }
+  }
+
+  Point buf_loc(x, y);
+  Instance* buffer
+      = resizer_->makeBuffer(buffer_cell, buffer_name.c_str(), parent, buf_loc);
+  inserted_buffer_count_++;
+
+  Pin* buffer_ip_pin = nullptr;
+  Pin* buffer_op_pin = nullptr;
+  resizer_->getBufferPins(buffer, buffer_ip_pin, buffer_op_pin);
+
+  // make sure any nets created are scoped within hierarchy
+  // backwards compatible. new naming only used for hierarchy code.
+  std::string net_name = db_network_->hasHierarchy()
+                             ? resizer_->makeUniqueNetName(parent)
+                             : resizer_->makeUniqueNetName();
+  Net* new_net = db_network_->makeNet(net_name.c_str(), parent);
+
+  Net* buffer_ip_net = nullptr;
+  Net* buffer_op_net = nullptr;
+
+  if (top_primary_input || !preserve_outputs) {
+    //
+    // Case 1
+    //------
+    // A primary input or do not preserve the outputs
+    //
+    // use orig net as buffer ip (keep primary input name exposed)
+    // use new net as buffer op (ok to use new name on op of buffer).
+    // move loads to op side (so might need to rename any hierarchical
+    // nets to avoid conflict of names with primary input net).
+    //
+    // record the driver pin modnet, if any
+    odb::dbModNet* driver_pin_mod_net = db_network_->hierNet(driver_pin);
+
+    //
     // Copy signal type to new net.
-    dbNet* out_net_db = db_network_->staToDb(out_net);
-    dbNet* in_net_db = db_network_->staToDb(in_net);
-    out_net_db->setSigType(in_net_db->getSigType());
+    //
+    dbNet* ip_net_db = load_db_net;
+    dbNet* op_net_db = db_network_->staToDb(new_net);
+    op_net_db->setSigType(ip_net_db->getSigType());
+    out_net = new_net;
+
+    buffer_op_net = new_net;
+    buffer_ip_net = db_network_->dbToSta(ip_net_db);
 
-    // Move load pins to out_net.
     for (const Pin* pin : load_pins) {
       Port* port = network_->port(pin);
       Instance* inst = network_->instance(pin);
-
-      // do not disconnect/reconnect don't touch instances
       if (resizer_->dontTouch(inst)) {
         continue;
       }
+      // preserve any hierarchical connection on the load
+      // & also connect the buffer output net to this pin
+      load_mod_net = db_network_->hierNet(pin);
+      load_db_net = db_network_->flatNet(pin);
+
       sta_->disconnectPin(const_cast<Pin*>(pin));
-      sta_->connectPin(inst, port, out_net);
+      sta_->connectPin(inst, port, buffer_op_net);
+
+      if (load_mod_net) {
+        db_network_->connectPin(const_cast<Pin*>(pin),
+                                db_network_->dbToSta(load_mod_net));
+      }
```

Comment:
Splitting connectPin into two API calls leaves room for error.  It would be better to take a possibly null mod_net in one call.